### ISSUE TITLE
[Usage]: [swift with vllm and only using vllm serve]  leads to different result（10% diff）

### ISSUE THREAD

**Issue by rainays** [vllm-project/vllm#16632]
### Your current environment

```text
The output of `python collect_env.py`
```
 **~ $ python collect_env.py**
INFO 04-15 11:43:18 __init__.py:190] Automatically detected platform cuda.
Collecting environment information...
/home/work/py39/lib/python3.9/site-packages/_distutils_hack/__init__.py:26: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")
PyTorch version: 2.5.1+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 12.1.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.17

Python version: 3.9.2 (default, Mar  3 2021, 20:02:32)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.10.0-1.0.0.34-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 10.1.105
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA A10
GPU 1: NVIDIA A10

Nvidia driver version: 535.154.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                256
On-line CPU(s) list:   0-255
Thread(s) per core:    2
Core(s) per socket:    64
Socket(s):             2
NUMA node(s):          2
Vendor ID:             AuthenticAMD
CPU family:            25
Model:                 1
Model name:            AMD EPYC 7W83 64-Core Processor
Stepping:              1
CPU MHz:               3200.910
CPU max MHz:           3673.0950
CPU min MHz:           1500.0000
BogoMIPS:              5090.40
Virtualization:        AMD-V
L1d cache:             32K
L1i cache:             32K
L2 cache:              512K
L3 cache:              32768K
NUMA node0 CPU(s):     0-63,128-191
NUMA node1 CPU(s):     64-127,192-255
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin brs arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif umip pku vaes vpclmulqdq rdpid overflow_recov succor smca fsrm

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] nvidia-cublas-cu11==11.11.3.6
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu11==11.8.87
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu11==11.8.89
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu11==11.8.89
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu11==9.1.0.70
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu11==10.9.0.58
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu11==10.3.0.86
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu11==11.4.1.48
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu11==11.7.5.86
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-ml-py==12.560.30
[pip3] nvidia-nccl-cu11==2.21.5
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu11==11.8.86
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] onnx==1.16.2
[pip3] onnxruntime==1.19.0
[pip3] pynvml==12.0.0
[pip3] pyzmq==26.2.0
[pip3] torch==2.5.1+cu118
[pip3] torchaudio==2.5.1
[pip3] torchvision==0.20.1+cu118
[pip3] transformers==4.49.0.dev0
[pip3] transformers-stream-generator==0.0.5
[pip3] triton==3.1.0
[conda] Could not collect
ROCM Version: Could not collect
Neuron SDK Version: N/A
vLLM Version: 0.7.2
vLLM Build Flags:
CUDA Archs: Not Set; ROCm: Disabled; Neuron: Disabled
GPU Topology:
	GPU0	GPU1	NIC0	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	SYS	SYS	0-63,128-191	0		N/A
GPU1	SYS	 X 	SYS	0-63,128-191	0		N/A
NIC0	SYS	SYS	 X

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_bond_0

CUDA_HOME=/home/work/cuda-11.8
CUDA_HOME=/home/work/cuda-11.8
LD_LIBRARY_PATH=/home/work/py39/lib/python3.9/site-packages/cv2/../../lib64:/home/work/py39/lib/python3.9/site-packages/nvidia/nvjitlink/lib:
MKL_DEBUG_CPU_TYPE=5
NCCL_CUMEM_ENABLE=0
TORCHINDUCTOR_COMPILE_THREADS=1
CUDA_MODULE_LOADING=LAZY

### How would you like to use vllm

my model base is qwen2.5vl-7b and we trained it with swift, and when i deploy with swift (vllm accelerate) and use vllm serve， it lead to different result
**The task is to determine whether the two signboards are the signboards of the same store.**
**swift deploy script：**
same Accuracy ： 0.9695121951219512 =159/164
same Recall 0.7871287128712872 =159/202
client：
`
/root/paddlejob/workspace/env_run/yinzilong/miniconda3_vllm/bin/swift deploy \
    --adapters /root/paddlejob/workspace/env_run/yinzilong/panchong/v3-20250312-205937/checkpoint-20000 \
    --infer_backend vllm \
    --temperature 0 \
    --gpu_memory_utilization 0.9 \
    --max_new_tokens 2048 \
    --tensor_parallel_size 1 \
    --logprobs true \
    --limit_mm_per_prompt '{"image": 2}' \
    --served_model_name 'qwen2_5_vl' \
    --host '10.96.206.107' \
    --port 8601
`
query：
`
def infer_batch(engine: 'InferEngine', infer_requests: List['InferRequest']):
    request_config = RequestConfig(max_tokens=512, temperature=0, logprobs=True, top_logprobs=2)
    metric = InferStats()
    resp_list = engine.infer(infer_requests, request_config, metrics=[metric])
    query0 = infer_requests[0].messages[0]['content']
    # print(f'query0: {query0}')
    # print(f'response0: {resp_list[0].choices[0].message.content}')
    # print(f'metric: {metric.compute()}')
    metric.reset()  # reuse
    return resp_list
if pic1_url == "" and pic2_url == "":
    infer_requests = [InferRequest(**get_data('text', query, []))]
elif pic1_url != "" and pic2_url == "":
    bid_image = get_pic_whole(pic1_url, pic1_ori_pos)
    infer_requests = [InferRequest(**get_data('image', query, [bid_image]))]
elif pic1_url == "" and pic2_url != "":
    bid_image = get_pic_whole(pic2_url, pic2_ori_pos)
    infer_requests = [InferRequest(**get_data('image', query, [bid_image]))]
else:
    bid_image1 = get_pic_whole(pic1_url, pic1_ori_pos)
    bid_image2 = get_pic_whole(pic2_url, pic2_ori_pos)
    infer_requests = [InferRequest(**get_data('image', query, [bid_image1, bid_image2]))]
resp_list = infer_batch(self.engine, infer_requests)
response_content = resp_list[0].choices[0].message.content
logprobs = resp_list[0].choices[0].logprobs
`
**when i use vllm**
same Accuracy： 0.8835978835978836 =167/189
same Recall: 0.8267326732673267 =167/202

client
`
VLLM_USE_V1=0 \
VLLM_WORKER_MULTIPROC_METHOD=spawn \
/root/paddlejob/workspace/env_run/yinzilong/miniconda3_vllm/bin/vllm serve \
/root/paddlejob/workspace/env_run/yinzilong/panchong/output/v3-20250312-205937/checkpoint-20000-merged \
--gpu-memory-utilization 0.9 \
--swap-space 8 \
--max-model-len 2048 \
--tensor-parallel-size 1 \
--host 10.95.192.17 \
--port 8601 \
--limit-mm-per-prompt "image=2" 
`
query with openai:
`
resp = self.client.chat.completions.create(model=self.model, 
                                                       messages=messages, 
                                                       max_tokens=10240, 
                                                       temperature=0,
                                                        logprobs=True,
                                                        top_logprobs=1)
`
it also happens：Compare whit swift request and swift deploy，**Requests are much slower**  when we use vllm serve and openai request.

Looking forward to your reply 🥺 thanks

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by DarkLight1337** [comment]
There was a bug in resolving Qwen2-VL chat template that was solved in a later version. Can you try updating vLLM and see if the problem persists?

**Comment by rainays** [comment]
> There was a bug in resolving Qwen2-VL chat template that was solved in a later version. Can you try updating vLLM and see if the problem persists?

whats bug？ becasue we can only use centos7, due to environmental constraints, we cannot upgrade vllm for the time being, any other solutions 🥺

**Comment by DarkLight1337** [comment]
https://github.com/vllm-project/vllm/pull/15143

**Comment by rainays** [comment]
> [#15143](https://github.com/vllm-project/vllm/pull/15143)

I also have a question 
when i use vllm serve and use openai  request

If there are multiple images, will the position of the images be determined according to the position of the placeholder in my text? Or will all the images be squeezed behind the text?

Because if my request refers to two images, my prompt is
poi1 name, address, coordinates, image, etc.
poi2 name, address, coordinates, image, etc.

If they are all squeezed together, I may not know which image belongs to

**Comment by DarkLight1337** [comment]
For Qwen2-VLm it should be based on the order of the `"image_url"` and `"text"` type messages you passed to OpenAI API

**Comment by yinzilong** [comment]
> For Qwen2-VL series it should be based on the order of the `"image_url"` and `"text"` type messages you passed to OpenAI API

Question:

My prompt:
poi1: name1, address1, coordinates1, image1<image>; poi2: name2, address2, coordinates2, image2<image>.
Provided image links:
image_url1
image_url2

I have provided both textual features and images for two POIs. Will the model understand that image1 belongs to poi1 and image2 belongs to poi2? Swift supports placeholders, but OpenAI doesn’t have the concept of placeholders.

My client-side message looks like this:

message = {
    'role': 'user',
    'content': [
        {
            'type': 'image_url',
            'image_url': {"url": f"data:image/png;base64,{image_base64_list[0]}"},
        }, {
            'type': 'image_url',
            'image_url': {"url": f"data:image/png;base64,{image_base64_list[1]}"},
        }, {
            'type': 'text',
            'text': query
        }
    ]
}
Will OpenAI simply concatenate the content as image, image, text, or will it respect the structure in the prompt like xxxxxx image1 xxxxxx image2 xxxx?

**Comment by yinzilong** [comment]
> For Qwen2-VL series it should be based on the order of the `"image_url"` and `"text"` type messages you passed to OpenAI API

hank you for your response. Let me refine this question further.

My prompt:
poi1: name1, address1, coordinates1, image1; poi2: name2, address2, coordinates2, image2.
Provided image links:
image_url1
image_url2

I have provided both textual features and images for two POIs. Will the model understand that image1 belongs to poi1 and image2 belongs to poi2? Swift supports placeholders, but OpenAI doesn’t have the concept of placeholders.

My client-side message looks like this:

message = {
'role': 'user',
'content': [
{
'type': 'image_url',
'image_url': {"url": f"data:image/png;base64,{image_base64_list[0]}"},
}, {
'type': 'image_url',
'image_url': {"url": f"data:image/png;base64,{image_base64_list[1]}"},
}, {
'type': 'text',
'text': query
}
]
}
Will OpenAI simply concatenate the content as image, image, text, or will it respect the structure in the prompt like xxxxxx image1 xxxxxx image2 xxxx?

---中文版本---
我的prompt: poi1: 名称1，地址1，坐标1，图片1；poi2:名称2，地址2，坐标2，图片2.
同时提供的图片:image_url1， image_url2

我既提供了两个poi的文本特征，也提供了两张图片。最终模型是否能知道图片1属于poi1，图片2属于poi2。swift 是支持占位符的，但是openai 没有占位符的概念。

我的客户端长这样，是否open会直接拼接为 图片，图片，文本。而不是这样  prompt 里面的 xxxxxx图片1xxxxx图片2xxx

message = {
‘role’: ‘user’,
‘content’: [
{
‘type’: ‘image_url’,
‘image_url’: {“url”: f"data:image/png;base64,{image_base64_list[0]}“},
}, {
‘type’: ‘image_url’,
‘image_url’: {“url”: f"data:image/png;base64,{image_base64_list[1]}”},
},{
‘type’: ‘text’,
‘text’: query
}]
}

**Comment by yinzilong** [comment]
> It will do this

Thank you for your reply. If the OpenAI client does not support this placeholder-based multi-image approach, are there other client-side request methods available? From what I currently understand, the Swift client can support multi-image with placeholders, but it does not support multimodal calls to the vLLM server.

**Comment by DarkLight1337** [comment]
Can you modify the list of messages? If your goal is to have "query1, image1, query2, image2", then you can set message like this:

```python
message = {
    ‘role’: ‘user’,
    ‘content’: [
        {‘type’: ‘text’, ‘text’: query1},
        {
            ‘type’: ‘image_url’,
            ‘image_url’: {“url”: f"data:image/png;base64,{image_base64_list[0]}“},
        },
        {‘type’: ‘text’, ‘text’: query2},
        {
                ‘type’: ‘image_url’,
                 ‘image_url’: {“url”: f"data:image/png;base64,{image_base64_list[1]}”},
        }
    ]
}
```

**Comment by yinzilong** [comment]
> Can you modify the list of messages? If your goal is to have "query1, image1, query2, image2", then you can set message like this:
> 
> message = {
>     ‘role’: ‘user’,
>     ‘content’: [
>         {‘type’: ‘text’, ‘text’: query1},
>         {
>             ‘type’: ‘image_url’,
>             ‘image_url’: {“url”: f"data:image/png;base64,{image_base64_list[0]}“},
>         },
>         {‘type’: ‘text’, ‘text’: query2},
>         {
>                 ‘type’: ‘image_url’,
>                  ‘image_url’: {“url”: f"data:image/png;base64,{image_base64_list[1]}”},
>         }
>     ]
> }

oh, yes, i try it, thank you
